gds_write/build_defs.bzl

```diff
@@ -0,0 +1,127 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""These build rules run automated GDS write on an implemented design"""
+
+load("//place_and_route:open_road.bzl", "OpenRoadInfo")
+load("//synthesis:build_defs.bzl", "SynthesisInfo")
+
+def _gds_write_impl(ctx):
+    final_gds = ctx.actions.declare_file("{}.gds".format(ctx.attr.name))
+    klayout_final_lyt = ctx.actions.declare_file("{}_klayout.lyt".format(ctx.attr.name))
+    klayout_final_lef = ctx.actions.declare_file("{}_klayout.lef".format(ctx.attr.name))
+    # Declare a list of additional LEF files to be copied into bazel build directory
+    final_lefs = []
+    for lef in ctx.files.additional_lef:
+        final_lefs.append(ctx.actions.declare_file("{}_{}.lef".format(ctx.attr.name, lef.basename.split('.', 1)[0])))
+
+    # Copy additional LEF files to bazel build directory
+    gen_lef_files_cmd = ""
+    for i in range(len(ctx.files.additional_lef)):
+        gen_lef_files_cmd += "cp {} {}; ".format(ctx.files.additional_lef[i].path, final_lefs[i].path)
+
+    ctx.actions.run_shell(
+        outputs = final_lefs,
+        inputs = ctx.files.additional_lef,
+        command = gen_lef_files_cmd,
+        mnemonic = "CopyAdditionalLefFiles"
+    )
+
+    # Copy KLayout LEF file to bazel build directory
+    ctx.actions.run_shell(
+        outputs = [klayout_final_lef],
+        inputs = [ctx.file.klayout_lef],
+        command = "cp {} {}; ".format(ctx.file.klayout_lef.path, klayout_final_lef.path),
+        mnemonic = "CopyKLayoutLefFile"
+    )
+
+    # Fix path to KLayout LEF file in KLayout technology file
+    ctx.actions.run_shell(
+        outputs = [
+            klayout_final_lyt
+        ],
+        inputs = depset(
+            [ctx.file.klayout_lyt, klayout_final_lef] +
+            final_lefs
+        ),
+        # Replace existing entry with reference to correct KLayout LEF file
+        # Save the results in klayout_final.lyt in bazel build directory
+        command =
+            "sed " +
+            "\"s/<lef-files>.*<\\/lef-files>/<lef-files>{}<\\/lef-files>/g\" ".format(klayout_final_lef.basename) +
+            "{} ".format(ctx.file.klayout_lyt.path) +
+            "> {}".format(klayout_final_lyt.path),
+        mnemonic = "FixPathInKlayoutTechFile"
+    )
+
+    # PYTHONPATH environment variable with python script imports
+    pythonpath = ""
+    for imp in ctx.attr._gds_write[PyInfo].imports.to_list():
+        pythonpath += ":{}".format(imp)
+
+    # Write GDS generation shell script
+    additional_gds_paths = " ".join([file.path for file in ctx.files.additional_gds])
+    write_gds_script = ctx.actions.declare_file("write_gds.sh")
+    write_gds_cmd = "PYTHONPATH=$PYTHONPATH:{} python {}".format(pythonpath, ctx.executable._gds_write.path) + \
```

Comment:
curious why we need an intermediate script rather than `run_shell`?

Reply:
Managed to change that to single `run_shell`